src/execution_plan/execution_plan.c

```diff
@@ -498,304 +482,360 @@ static void _NewExecutionPlanSegment(RedisModuleCtx *ctx, GraphContext *gc,
+
+	uint skip = 0;
+	uint limit = RESULTSET_UNLIMITED;
+	if(skip_clause) skip = AST_ParseIntegerNode(skip_clause);
+	if(limit_clause) limit = AST_ParseIntegerNode(limit_clause);
+
+	int sort_direction = 0;
+	if(segment->order_expressions) {
+		const cypher_astnode_t *order_clause = cypher_ast_return_get_order_by(clause);
+		sort_direction = AST_PrepareSortOp(order_clause);
 	}
 
-	assert(!(with_clause && ret_clause));
+	_buildProjectionOps(segment, skip, limit, sort_direction, aggregate, distinct, ops);
+}
 
-	uint *modifies = NULL;
+static void _buildWithOps(ExecutionPlanSegment *segment, const cypher_astnode_t *clause,
+						  Vector *ops) {
+	bool aggregate = AST_ClauseContainsAggregation(clause);
+	bool distinct = cypher_ast_with_is_distinct(clause);
 
-	// WITH/RETURN projections have already been constructed from the AST
-	AR_ExpNode **projections = segment->projections;
+	const cypher_astnode_t *skip_clause = cypher_ast_with_get_skip(clause);
+	const cypher_astnode_t *limit_clause = cypher_ast_with_get_limit(clause);
 
-	if(with_clause || ret_clause || call_clause) {
-		// TODO improve interface, maybe CollectEntityIDs variant that builds an array
-		rax *modifies_ids = raxNew();
-		uint exp_count = array_len(projections);
-		for(uint i = 0; i < exp_count; i ++) {
-			AR_ExpNode *exp = projections[i];
-			AR_EXP_CollectEntityIDs(exp, modifies_ids);
-		}
+	uint skip = 0;
+	uint limit = RESULTSET_UNLIMITED;
+	if(skip_clause) skip = AST_ParseIntegerNode(skip_clause);
+	if(limit_clause) limit = AST_ParseIntegerNode(limit_clause);
 
-		modifies = array_new(uint, raxSize(modifies_ids));
-		raxIterator iter;
-		raxStart(&iter, modifies_ids);
-		raxSeek(&iter, ">=", (unsigned char *)"", 0);
-		while(raxNext(&iter)) {
-			modifies = array_append(modifies, *(uint *)iter.key);
-		}
-		raxFree(modifies_ids);
+	int sort_direction = 0;
+	if(segment->order_expressions) {
+		const cypher_astnode_t *order_clause = cypher_ast_with_get_order_by(clause);
+		sort_direction = AST_PrepareSortOp(order_clause);
 	}
 
-	OpBase *op;
+	_buildProjectionOps(segment, skip, limit, sort_direction, aggregate, distinct, ops);
+}
 
-	if(with_clause) {
-		// uint *with_projections = AST_WithClauseModifies(ast, with_clause);
-		if(AST_ClauseContainsAggregation(with_clause)) {
-			op = NewAggregateOp(projections, modifies);
-		} else {
-			op = NewProjectOp(projections, modifies);
-		}
-		Vector_Push(ops, op);
+static void _ExecutionPlanSegment_ConvertClause(GraphContext *gc, AST *ast,
+												ExecutionPlanSegment *segment, ResultSetStatistics *stats, const cypher_astnode_t *clause,
+												Vector *ops) {
+	OpBase *op = NULL;
+	cypher_astnode_type_t t = cypher_astnode_type(clause);
+	// Because 't' is set using the offsetof() call, it cannot be used in switch statements.
+	if(t == CYPHER_AST_MATCH || t == CYPHER_AST_CALL) {
+		// MATCH and CALL clauses are processed before looping over all clauses.
```

Comment:
What's the plan here, how difficult do you think it will be to process MATHC/CALL clauses here?

Reply:
MATCH will, I think, be fairly tricky, because right now we use all MATCH clauses to form one QueryGraph, then from that any number of scans, traversals, and even cartesian products. I think changing that logic will entail moving away from a monolithic QueryGraph (or at least associating different clauses with different paths in that graph). I do think this will be eventually necessary for MERGE support, OPTIONAL MATCH, and possibly WHERE paths.

CALL on the other hand is not too hard; it just introduces some problems right now because implicit YIELDs introduce projections that are not seen in the AST.